posts/inside-rust/2023-05-03-stabilizing-async-fn-in-trait.md

```diff
@@ -0,0 +1,229 @@
+}
+```
+
+Even though the need for "impl trait in traits" comes up a lot in async, they are a general feature that will be useful in many contexts having nothing to do with async (for example, returning iterators from trait methods).
+
+**Status:** Return-position impl trait in traits have an experimental implementation and are described in the recently opened [RFC 3425].
+
+[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=75cfc199cc50a111576c2d8e342ae823)
+
+[RFC 3425]: https://github.com/rust-lang/rfcs/pull/3425
+
+## Evaluating the MVP
+
+To evaluate the utility of this MVP, the working group collected [five case studies] covering the [builder-provider pattern used in the AWS SDK](https://rust-lang.github.io/async-fundamentals-initiative/evaluation/case-studies/builder-provider-api.html#dynamic-dispatch-behind-the-api); the potential use of async function in traits in [tower][cst] and the actual use in [embassy][cse], the [Fuchsia networking stack] and [an internal Microsoft tool]. These studies validated that the above functionality is sufficient to use async function in traits for all kinds of things, though some situations require workarounds (hence the "MVP" title).
+
+[Fuchsia networking stack]: https://rust-lang.github.io/async-fundamentals-initiative/evaluation/case-studies/socket-handler.html
+
+[an internal Microsoft tool]: https://rust-lang.github.io/async-fundamentals-initiative/evaluation/case-studies/microsoft.html
+
+[cst]: https://rust-lang.github.io/async-fundamentals-initiative/evaluation/case-studies/tower.html
+
+[cse]: https://rust-lang.github.io/async-fundamentals-initiative/evaluation/case-studies/embassy.html
+
+[five case studies]: https://rust-lang.github.io/async-fundamentals-initiative/evaluation/case-studies.html
+
+## What the MVP will not support or won't support well
+
+The case studies revealed two situations that the MVP doesn't support very well, but both of them have workarounds available. These workarounds are mechanical and once the MVP is available on stable it will be possible to automate them via a custom derive or other crates on crates.io.
+
+### Modeling dynamic dispatch
+
+In the MVP, traits that use async functions are not "dyn safe", meaning that they don't support dynamic dispatch. So e.g. given the `HealthCheck` trait we saw earlier, one could not write `Box<dyn HealthCheck>`.
+
+At first, this seems like a crucial limitation, since many of the use cases require dynamic dispatch! But it turns out that there is a workaround. One can define an "erased" trait internally to your crate that enables dynamic dispatch. The process was pioneered by crates like [erased serde] and is explained in detail in the [builder-provider case study].
+
+To make this workaround easier in the near term, we hope to provide a proc macro to automate it. In the future, async fn [should work][ugf-dyn] with `dyn Trait` directly.
```

Comment:
```suggestion
To make this workaround easier in the near term, we plan to provide a proc macro to automate it. In the future, async fn [should work][ugf-dyn] with `dyn Trait` directly.
```

I'd like to make this sound like are committing to delivering this proc macro quickly, rather than just hoping we or someone else writes the macro.